fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java

```diff
@@ -257,6 +283,193 @@ public PlanFragment visitPhysicalAggregate(
+         * prePressions: k1, k2
+         */
+        // create virtual tupleDesc
+        List<VirtualSlotReference> virtualSlotList = virtualGroupByExpressions.stream()
+                .filter(VirtualSlotReference.class::isInstance)
+                .map(VirtualSlotReference.class::cast)
+                .collect(Collectors.toList());
+        List<Slot> virtualSlots = Lists.newArrayList(virtualSlotList);
+        TupleDescriptor virtualTupleDesc = generateTupleDesc(virtualSlots, null, context);
+
+        // create repeat slots and PreExpressions
+        List<SlotReference> groupBySlot = genGroupBySlotList(groupByExpressions, outputExpressionList);
+        List<Slot> finalSlots = genOutputSlots(groupBySlot, outputExpressionList, virtualSlots);
+
+        List<Expression> preExpressions = genPreExpressions(nonVirtualGroupByExpressions, outputExpressionList);
+        ArrayList<Expr> preRepeatExprs = preExpressions.stream()
+                .map(e -> ExpressionTranslator.translate(e, context)).collect(Collectors.toCollection(ArrayList::new));
+
+        // create output TupleDesc
+        TupleResult tupleResult = genTupleDescAndDescList(finalSlots, null, context);
+        TupleDescriptor outputTupleDesc = tupleResult.getTupleDescriptor();
+        setSlotNullable(outputTupleDesc, finalSlots);
+
+        GroupingInfo groupingInfo = new GroupingInfo(ExpressionTranslator.translateGroupingType(repeat),
+                virtualTupleDesc, outputTupleDesc, preRepeatExprs);
+
+        // 2.create repeat node
+        // Replace the bitset of groupingsets with the bitset corresponding to slotId
+        List<Set<Integer>> repeatSlotIdList = genGroupingIdList(
+                repeat.useBitsetsToRepresentGroupingSets(), tupleResult.getSlotDescriptors());
+        Set<Integer> allSlotId = genAllSlotId(repeatSlotIdList);
+
+        RepeatNode repeatNode = new RepeatNode(context.nextPlanNodeId(),
+                inputPlanFragment.getPlanRoot(), groupingInfo, repeatSlotIdList,
+                allSlotId, repeat.getVirtualSlotValues());
+        repeatNode.setNumInstances(inputPlanFragment.getPlanRoot().getNumInstances());
+        inputPlanFragment.addPlanRoot(repeatNode);
+        inputPlanFragment.updateDataPartition(DataPartition.RANDOM);
+        return inputPlanFragment;
+    }
+
+    /**
+     * Generate outputSlot based on groupBy and output information.
+     * The generated outputSlot needs to guarantee the order,
+     * first slotReference and then virtualSlotReference.
+     *
+     * First collect the slot and then collect the virtualSlot.
+     * 1. Put the columns in groupBy into the slot
+     * 2. Put the columns in the output into the slot
+     * 3. Put slot and virtualSlot into finalSLots in turn.
+     *
+     * eg: select sum(k2) grouping(k1) from t1 group by grouping sets((k1));
+     * 1. slots:[k1]
+     * 2. slots:[k1, k2]
+     * 3. finalSlots: [k1, k2, GROUPING_PREFIX_k1]
+     */
+    private List<Slot> genOutputSlots(
```

Comment:
output maybe not slot, but contains slot, and you need sort it, so `getUsedSortedSlots` is more reasonable. And primary method should move to the tailing of the class.

And this method should be simplified to
```java
private List<Slot> getUsedSortedSlots(List<Expression> groupByExpressions, List<NamedExpression> outputExpressions)
    Set <Slot> allSlots = ImmutableSet.builder()
        .addAll(groupByExpressions.stream()
            .flatMap(output -> output.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        .addAll(outputExpressions.stream()
            .flatMap(output -> output.getInputSlots().stream()))
            .collect(ImmutableSet.toSet())
        ).build();

    Set<Slot> nonVirtualSlots = allSlots.stream()
        .filter(slot -> !(slot instanceof VirtualSlotReference)
        .collect(ImmutableSet.toSet());

    Set<Slot> virtualSlots = allSlots.stream()
        .filter(slot -> !(slot instanceof VirtualSlotReference)
        .collect(ImmutableSet.toSet());

    return ImmutableList.Builder()
        .addAll(nonVirtualSlots)
        .addAll(virtualSlots);
}
```